### PR TITLE
ROX-36250: implement node report service API

### PR DESCRIPTION
**Pull request by janisz**
This commit implements the complete CRUD service for node vulnerability
reports, following StackRox v2 API patterns and security best practices.

Service operations:
- GetNodeReportConfigurations: List all node report configs with pagination
- GetNodeReportConfiguration: Get single config by ID
- GetNodeReportHistory: Query report execution history with filtering
- CreateNodeReportConfiguration: Create new scheduled/on-demand reports
- UpdateNodeReportConfiguration: Update existing configurations
- DeleteNodeReportConfiguration: Delete configs (requires WorkflowAdmin)
- RunNodeReport: Execute on-demand reports
- CancelNodeReport: Cancel running reports (requires WorkflowAdmin)
- DownloadNodeReport: Retrieve generated report files

Security and authorization:
- Enforce View(Node) permission for read operations
- Require View(WorkflowAdministration) for delete/cancel operations
- Validate report configurations before persistence
- Scope queries by report type to prevent data leakage

Bug fixes:
- Add ReportType filter to history queries (prevent mixing image/node reports)
- Fix blob path for view-based reports (use "view-based-report" directory)
- Correct permissions on delete/cancel endpoints

Implementation details:
- Use errox pattern for consistent error handling
- Integrate with report scheduler for execution
- Support both scheduled and on-demand report generation
- Handle report cancellation with proper state management
- Blob storage integration for report file management

Testing:
- Comprehensive unit tests for all CRUD operations
- PostgreSQL integration tests with real database
- Test error paths and edge cases
- Validate authorization enforcement

Co-Authored-By: Claude Sonnet 4.5 <noreply@anthropic.com>

## Description

change me!

## User-facing documentation

- [ ] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [ ] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [ ] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

### How I validated my change

change me!

### CODE REVIEW

**Comment by janisz** [comment]
This change is part of the following stack:

- #22420
    - #22505
        - #22506
            - #22507
                - #22508 ◀
                    - #22388

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: central/reports/service/v2/node/download.go
Comment (charmik-redhat) [inline central/reports/service/v2/node/download.go:1]: I think the existing download handler for Image vulnerability reports can be reused here without any changes required other than the one I added in #22420 

https://github.com/stackrox/stackrox/pull/22420/changes#diff-51ba6d9518974624bbca58ff7a0d1a659a87aee96308f31581d25364f30a7b85R113

File: central/reports/service/v2/node/service_impl.go [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+	maxPaginationLimit = 1000
+)
+
+var (
+	log = logging.LoggerForModule()
+
+	workflowSAC = sac.ForResource(resources.WorkflowAdministration)
+
+	authorizer = perrpc.FromMap(map[authz.Authorizer][]string{
+		user.With(permissions.View(resources.WorkflowAdministration), permissions.View(resources.Node)): {
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go]: Add View cluster to all APIs. That was a decision we made during the reporting sync.

File: central/reports/service/v2/node/service_impl.go [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+	if !exists {
+		return nil, errox.NotFound.CausedByf("report configuration with id '%s' does not exist", req.GetId())
+	}
+
+	if config.GetType() != storage.ReportConfiguration_NODE_VULNERABILITY {
+		return nil, errox.InvalidArgs.CausedByf("report configuration '%s' is not a node vulnerability report", req.GetId())
+	}
+
+	if !common.HasValidResourceScope(config.GetResourceScope()) {
+		return nil, errox.InvalidArgs.CausedByf("report configuration '%s' has an empty resource scope (no collection ID or entity scope)", req.GetId())
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go]: nit: Remove "no collection ID" from error message as collection scoping is not allowed for node cve reports.

File: central/reports/service/v2/node/service_impl.go [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+	cancelled, err := s.scheduler.CancelReportRequest(ctx, req.GetId())
+	if err != nil {
+		return nil, err
+	}
+	if !cancelled {
+		reportStatus.RunState = storage.ReportStatus_FAILURE
+		reportStatus.ErrorMsg = "Job cancelled by user"
+		reportStatus.CompletedAt = protocompat.TimestampNow()
+		snapshot.ReportStatus = reportStatus
+		if err := s.snapshotDatastore.UpdateReportSnapshot(ctx, snapshot); err != nil {
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go]: Not cancelled should just return an error. If cancel did succeed, the snapshot with the FAILURE status and an error message is persisted by the scheduler. So we don't need to do that here.

File: central/reports/service/v2/node/service_impl.go [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+		return nil, errox.ServerError.CausedByf("scheduler error: %s", err)
+	}
+
+	return &apiV2.RunReportResponseViewBased{ReportID: reportID, RequestName: reportRequest.ReportSnapshot.GetName()}, nil
+}
+
+func (s *serviceImpl) GetViewBasedNodeReportHistory(ctx context.Context, req *apiV2.GetViewBasedReportHistoryRequest) (*apiV2.ReportHistoryResponse, error) {
+	queryBuilder := search.NewQueryBuilder().
+		AddExactMatches(search.ReportType, storage.ReportSnapshot_NODE_VULNERABILITY.String()).
+		AddStrings(search.ReportConfigID, "")
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go]: nit: We can search using ReportType = VIEW_BASED instead. The new ReportType column added to snapshots will make it possible.

File: central/reports/service/v2/node/service_impl.go [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+
+func (s *serviceImpl) GetViewBasedMyNodeReportHistory(ctx context.Context, req *apiV2.GetViewBasedReportHistoryRequest) (*apiV2.ReportHistoryResponse, error) {
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errors.New("could not determine user identity from provided context")
+	}
+
+	queryBuilder := search.NewQueryBuilder().
+		AddExactMatches(search.ReportType, storage.ReportSnapshot_NODE_VULNERABILITY.String()).
+		AddStrings(search.ReportConfigID, "")
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go]: nit: We can search using ReportType = VIEW_BASED instead

**Review by charmik-redhat — APPROVED** [review]

**Review by charmik-redhat — COMMENTED** [review]
Looks like this change for delegating report generation to central-worker pod got merged recently
https://github.com/stackrox/stackrox/pull/21572/changes#diff-f46fb9453c2ee696d12b7b78578b9114de2e7eef8fc337cab6f0cbd2d5e0af34

Can we update node report service to add those `pg_notify` calls too? We should only need the service changes. Other things like validation, handling of notifications, etc seem to be NODE_VULNERABILITY report aware. Can do it as followup PR too if you want to merge this one with clean CI and all.

File: central/reports/service/v2/node/service_impl.go [hunk 6]
```diff
@@ -1,23 +1,617 @@
...
+
+func userFromContext(ctx context.Context) (*storage.SlimUser, error) {
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errox.NoCredentials.New("could not determine user identity from provided context")
+	}
+	return slimUser, nil
+}
+
+var _ Service = (*serviceImpl)(nil)
```
Comment (charmik-redhat) [inline central/reports/service/v2/node/service_impl.go:617]: What does this do?

**Review by charmik-redhat — COMMENTED** [review]
Just one question. LGTM otherwise

File: central/reports/service/v2/node/service_impl.go (hunk 6 above)
Comment (janisz) [reply central/reports/service/v2/node/service_impl.go:617]: It's a compilation level check for the interface implementation, I'm not sure if it's need but we have it in some places and before wiring the service it's good to have to avoid silly mistakes of not implementing a method.

Comment (janisz) [reply central/reports/service/v2/node/service_impl.go:617]: https://github.com/search?q=repo%3Astackrox%2Fstackrox+%22var+_%22+path%3A%2F%5Ecentral%5C%2F%2F+path%3A%2F%5Ecentral%5C%2F%2F&type=code
